### PR TITLE
Add !-postfix to modifying methods, prefer to extends Base methods.

### LINKED ISSUE #161: Rename the setters in the `Window` API functions with a `!`?
Can I rename the modifying versions of the `Window` API functions with a `!` at the end?

That is, for these ones:
https://github.com/JunoLab/Blink.jl/blob/d984c4d99586c451395cba55eadf8b47d9cac091/src/AtomShell/window.jl#L87-L136

There's a setter and a getter for a bunch of them (`position`, `size`, etc).

Can I rename those to `position!(w, x, y)` and `position(w)`?

(I came across this as part of https://github.com/JunoLab/Blink.jl/issues/160 😄)

### PR DESCRIPTION
**Pull request by twavv**
This should™ be fully backwards compatible.

I didn't want to add deprecation messages because that would just be cluttering.

Closes #161.

### CODE REVIEW

**Comment by travigd** [comment]
Blocked by #233.

The issue is that the `size` method is ambiguous with one that was deprecated between Julia 0.6 and 0.7. There are a couple of solutions.
* Change the test to use `resize!` which is unambiguous (this would mean size is broken on Julia 0.7 but, again, not many people use Julia 0.7).
* Drop support for Julia 0.7 (this requires a non-patch release).
* Add a more specific method for `size(::Window, ::Integer, ::Integer)`.

Any preferences? :^)

**Comment by NHDaly** [comment]
> Add a more specific method for `size(::Window, ::Integer, ::Integer)`.

Yeah, i think this is the sanest thing to do. It's just keeping the existing overload that was already there. Mark it deprecatd, with a note that it's only there until we drop 0.7 support, something like:
```
# deprecated; delete once we drop julia 0.7 support
```
And then drop 0.7 in a non-patch release, later. :)

Does that sound good to you?

**Comment by NHDaly** [comment]
Question: can we just drop the old methods and make a non-patch release? Do we need to keep things backwards compatible?

Or do you think that should wait for something like a `1.0.0` release? Are you preparing for that now? 😁

File: src/AtomShell/window.jl [hunk 1]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
...
+    progress!(win::Window, p=-1)
 
 Sets progress value in progress bar. Valid range is [0, 1.0]. Remove progress
 bar when progress < 0; Change to indeterminate mode when progress > 1.
 
 https://github.com/electron/electron/blob/master/docs/api/browser-window.md#winsetprogressbarprogress-options
 """
-progress(win::Window, p = -1) =
-  @dot_ win setProgressBar($p)
+progress!(win::Window, p) = @dot_ win setProgressBar($p)
```
Comment (NHDaly) [inline src/AtomShell/window.jl]:
```suggestion
progress!(win::Window, p=-1) = @dot_ win setProgressBar($p)
```
Or did you drop the default value on purpose? If so, can you also drop it from the docstring?

File: src/AtomShell/window.jl [hunk 2]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
   @dot_ win flashFrame($on)
 
 """
-    progress(win::Window, p=-1)
+    progress!(win::Window, p=-1)
```
Comment (NHDaly) [inline src/AtomShell/window.jl]: It looks like you changed `win::Window` to `window` inconsistently in the following docstrings. Would you be able to go back and make them all consistent? Sorry. 😬 

I'm fine with either; up to you.  I see the appeal of `win::Window`, though: it really does work for any other types, and these aren't meant to be generic functions.

File: src/AtomShell/window.jl [hunk 3]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
...
+title(win::Window) = @dot win getTitle()
+
+"""
+    center!(win::Window)
+
+Center a window.
+"""
+center!(win::Window) = @dot_ win center()
 
+# Deprecated (and misspelled?)
```
Comment (NHDaly) [inline src/AtomShell/window.jl]: haha 😛 Maybe this was written by a british english speaker? Perhaps Shashi? I know he wrote a lot of this originally, right?

**Comment by travigd** [comment]
Moreso the latter. There's two things I'm thinking about.

* We might require more API changes, in which case, we'd want to limit breaking stuff to one release (hopefully the 1.0.0 release).
* We should release breaking changes with other improvements to increase buy-in (so increased stability, hopefully lower time-to-first-window)

File: src/AtomShell/window.jl [hunk 4]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
...
 
-size(win::Window, w::Integer, h::Integer) =
-  invoke(size, Tuple{Window, Any, Any}, win, w, h)
+# Deprecated
+Base.size(window::Window, width, height) = resize!(window, width, height)
 
-size(win::Window, w, h) =
-  @dot_ win setSize($w, $h)
+"""
+    size(window::$(repr(Window)))
```
Comment (NHDaly) [inline src/AtomShell/window.jl]: haha same here. I guess this was an attempt to generate `Blink.Window` or `MyPackage.Blink.Window` or w/e? Seems cool, i haven't seen anyone do that before

File: src/AtomShell/window.jl [hunk 5]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
...
 floating(win::Window, flag) =
   @dot_ win setAlwaysOnTop($flag)
 
 floating(win::Window) =
   @dot win isAlwaysOnTop()
 
-loadurl(win::Window, url) =
-  @dot win loadURL($url)
+loadurl!(win::Window, url) = @dot win loadURL($url)
+loadurl(win::Window, url) = loadurl!(win, url)
```
Comment (NHDaly) [inline src/AtomShell/window.jl]:
```suggestion
# deprecated
loadurl(win::Window, url) = loadurl!(win, url)
```

File: src/AtomShell/window.jl [hunk 6]
```diff
@@ -154,71 +163,116 @@ flashframe(win::Window, on = true) =
...
 
-loadfile(win::Window, f) =
-  loadurl(win, "file://$f")
+loadfile!(win::Window, f) = loadurl(win, "file://$f")
+loadfile(win::Window, f) = loadfile!(win, f)
```
Comment (NHDaly) [inline src/AtomShell/window.jl]:
```suggestion
# deprecated
loadfile(win::Window, f) = loadfile!(win, f)
```

**Comment by NHDaly** [comment]
Cool, yeah, those are good points! :) Love it.

Thanks for tackling this!

File: src/AtomShell/window.jl (hunk 4 above)
Comment (travigd) [reply src/AtomShell/window.jl]: I think it was but in hindsight I think it's just gonna be best to stick to `::Window` and maybe use DocStringExtensions at some point. I'm just gonna use `window::Window` to be consistent. :^)

**Comment by travigd** [comment]
Think I resolved most of your comments, will give another look.

One think I'd like some input on is whether or not we should rename some methods to use underscores instead of smooshing (e.g. `flash_frame!` instead of `flash_frame!`). [BlueStyle](https://github.com/invenia/BlueStyle) recommends underscores for non-Julia-Base code and I'm inclined to agree for consistency reasons (e.g. it's hard to find the line in the sand about where to add underscores).

The reason for this was that I was struggling to find a name for `setAlwaysOnTop` (this is a TODO in the PR). Options are `pin!` or `always_on_top!`. The former punts the underscore problem but 🤷‍♀️

**Comment by NHDaly** [comment]
Yeah, i think adding underscores is fine. I don't really have much preference, but my opinion leans more towards underscores than not. For example, even `load_file` seems preferable to `loadfile` (although `loadfile` is certainly cuter 😸)

**Pull request closed without merging**